crates/clvm-utils/src/tree_hash.rs

```diff
@@ -61,6 +60,106 @@ impl Deref for TreeHash {
     }
 }
 
+#[derive(Default)]
+pub struct TreeCache {
+    hashes: Vec<TreeHash>,
+    // each entry is an index into hashes, or u16::MAX if the pair has not been
+    // cached
+    pairs: Vec<u16>,
+}
+
+const NOT_VISITED: u16 = u16::MAX;
+#[allow(dead_code)]
+const SEEN_ONCE: u16 = u16::MAX - 1;
+const SEEN_MULTIPLE: u16 = u16::MAX - 2;
+
+impl TreeCache {
+    pub fn get(&self, n: NodePtr) -> Option<&TreeHash> {
+        // We only cache pairs (for now)
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return None;
+        }
+
+        let idx = n.index() as usize;
+        if idx >= self.pairs.len() {
+            return None;
+        }
+
+        let slot = self.pairs[idx];
+        if slot >= SEEN_MULTIPLE {
+            return None;
+        }
+        Some(&self.hashes[slot as usize])
+    }
+
+    pub fn insert(&mut self, n: NodePtr, hash: &TreeHash) {
+        // If we've reached the max size, just ignore new cache items
+        if self.hashes.len() == SEEN_MULTIPLE as usize {
+            return;
+        }
+
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return;
+        }
+
+        let idx = n.index() as usize;
+        if idx >= self.pairs.len() {
+            self.pairs.resize(idx + 1, NOT_VISITED);
+        }
+
+        let slot = self.hashes.len();
+        self.hashes.push(*hash);
+        self.pairs[idx] = slot as u16;
+    }
+
+    /// mark the node as being visited. Returns true if this is node has already
+    /// been visited. i.e. returns whether this node is a good candidate for
+    /// memoizing tree hash results.
+    fn visit(&mut self, n: NodePtr) -> bool {
+        if !matches!(n.object_type(), ObjectType::Pair) {
+            return false;
+        }
+        let idx = n.index() as usize;
+        if idx >= self.pairs.len() {
+            self.pairs.resize(idx + 1, NOT_VISITED);
```

Comment:
Should this return `false` if `self.hashes.len() == SEEN_MULTIPLE as usize` like in `insert`?

Reply:
good point, probably

Reply:
actually, it's the `hashes` Vec whose size is limited. `pairs` is capped by the `Allocator` limits.

Reply:
the way the cache is used is to first visit the tree, *then* to compute the tree hash. `insert()` is called during tree-hash computation. So, if we've  run the tree-hash, and the cache is full, you're saying there's no point in visiting any more nodes because nothing more will be cached anywy?